### ISSUE TITLE
pip install : ModuleNotFoundError: No module named 'keras.objectives'

### ISSUE THREAD

**Issue by AlexCalderwood** [theislab/dca#48]
I'm struggling to get the pip install to work. 
After 
`$pip install dca`
`$python`
`>>> from dca.api import dca`

I get the error 

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/calderwa/opt/anaconda3/envs/test/lib/python3.6/site-packages/dca/api.py", line 15, in <module>
    from .train import train
  File "/Users/calderwa/opt/anaconda3/envs/test/lib/python3.6/site-packages/dca/train.py", line 24, in <module>
    from .network import AE_types
  File "/Users/calderwa/opt/anaconda3/envs/test/lib/python3.6/site-packages/dca/network.py", line 27, in <module>
    from keras.objectives import mean_squared_error
ModuleNotFoundError: No module named 'keras.objectives'

I find that with the bioconda install I get the same error message.

Please see below for the full output from the pip install step:

test ❯ pip install dca
Collecting dca
  Using cached DCA-0.3.2-py3-none-any.whl (26 kB)
Collecting scikit-learn
  Using cached scikit_learn-0.24.2-cp36-cp36m-macosx_10_13_x86_64.whl (7.2 MB)
Collecting six>=1.10.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting keras>=2.4
  Using cached keras-2.6.0-py2.py3-none-any.whl (1.3 MB)
Collecting pandas
  Using cached pandas-1.1.5-cp36-cp36m-macosx_10_9_x86_64.whl (10.2 MB)
Collecting tensorflow>=2.0
  Using cached tensorflow-2.6.0-cp36-cp36m-macosx_10_11_x86_64.whl (198.9 MB)
Collecting numpy>=1.7
  Using cached numpy-1.19.5-cp36-cp36m-macosx_10_9_x86_64.whl (15.6 MB)
Collecting kopt
  Using cached kopt-0.1.0-py2.py3-none-any.whl
Collecting scanpy
  Using cached scanpy-1.7.2-py3-none-any.whl (10.3 MB)
Collecting h5py
  Using cached h5py-3.1.0-cp36-cp36m-macosx_10_9_x86_64.whl (2.9 MB)
Collecting flatbuffers~=1.12.0
  Using cached flatbuffers-1.12-py2.py3-none-any.whl (15 kB)
Collecting termcolor~=1.1.0
  Using cached termcolor-1.1.0-py3-none-any.whl
Collecting protobuf>=3.9.2
  Downloading protobuf-3.18.0-cp36-cp36m-macosx_10_9_x86_64.whl (1.0 MB)
     |████████████████████████████████| 1.0 MB 3.4 MB/s
Collecting google-pasta~=0.2
  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)
Collecting astunparse~=1.6.3
  Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
Collecting grpcio<2.0,>=1.37.0
  Using cached grpcio-1.40.0-cp36-cp36m-macosx_10_10_x86_64.whl (3.9 MB)
Collecting absl-py~=0.10
  Using cached absl_py-0.13.0-py3-none-any.whl (132 kB)
Collecting tensorflow-estimator~=2.6
  Using cached tensorflow_estimator-2.6.0-py2.py3-none-any.whl (462 kB)
Collecting opt-einsum~=3.3.0
  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)
Collecting gast==0.4.0
  Using cached gast-0.4.0-py3-none-any.whl (9.8 kB)
Collecting typing-extensions~=3.7.4
  Using cached typing_extensions-3.7.4.3-py3-none-any.whl (22 kB)
Collecting six>=1.10.0
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting wrapt~=1.12.1
  Using cached wrapt-1.12.1-cp36-cp36m-macosx_10_9_x86_64.whl
Collecting tensorboard~=2.6
  Using cached tensorboard-2.6.0-py3-none-any.whl (5.6 MB)
Collecting keras-preprocessing~=1.1.2
  Using cached Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)
Requirement already satisfied: wheel~=0.35 in ./opt/anaconda3/envs/test/lib/python3.6/site-packages (from tensorflow>=2.0->dca) (0.37.0)
Collecting clang~=5.0
  Using cached clang-5.0-py3-none-any.whl
Collecting cached-property
  Using cached cached_property-1.5.2-py2.py3-none-any.whl (7.6 kB)
Requirement already satisfied: setuptools>=41.0.0 in ./opt/anaconda3/envs/test/lib/python3.6/site-packages (from tensorboard~=2.6->tensorflow>=2.0->dca) (52.0.0.post20210125)
Collecting markdown>=2.6.8
  Using cached Markdown-3.3.4-py3-none-any.whl (97 kB)
Collecting google-auth<2,>=1.6.3
  Using cached google_auth-1.35.0-py2.py3-none-any.whl (152 kB)
Collecting requests<3,>=2.21.0
  Using cached requests-2.26.0-py2.py3-none-any.whl (62 kB)
Collecting tensorboard-plugin-wit>=1.6.0
  Using cached tensorboard_plugin_wit-1.8.0-py3-none-any.whl (781 kB)
Collecting werkzeug>=0.11.15
  Using cached Werkzeug-2.0.1-py3-none-any.whl (288 kB)
Collecting google-auth-oauthlib<0.5,>=0.4.1
  Using cached google_auth_oauthlib-0.4.6-py2.py3-none-any.whl (18 kB)
Collecting tensorboard-data-server<0.7.0,>=0.6.0
  Using cached tensorboard_data_server-0.6.1-py3-none-macosx_10_9_x86_64.whl (3.5 MB)
Collecting cachetools<5.0,>=2.0.0
  Using cached cachetools-4.2.2-py3-none-any.whl (11 kB)
Collecting pyasn1-modules>=0.2.1
  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
Collecting rsa<5,>=3.1.4
  Using cached rsa-4.7.2-py3-none-any.whl (34 kB)
Collecting requests-oauthlib>=0.7.0
  Using cached requests_oauthlib-1.3.0-py2.py3-none-any.whl (23 kB)
Collecting importlib-metadata
  Using cached importlib_metadata-4.8.1-py3-none-any.whl (17 kB)
Collecting pyasn1<0.5.0,>=0.4.6
  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
Collecting charset-normalizer~=2.0.0
  Using cached charset_normalizer-2.0.5-py3-none-any.whl (37 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.6-py2.py3-none-any.whl (138 kB)
Requirement already satisfied: certifi>=2017.4.17 in ./opt/anaconda3/envs/test/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow>=2.0->dca) (2021.5.30)
Collecting idna<4,>=2.5
  Using cached idna-3.2-py3-none-any.whl (59 kB)
Collecting oauthlib>=3.0.0
  Using cached oauthlib-3.1.1-py2.py3-none-any.whl (146 kB)
Collecting dataclasses
  Using cached dataclasses-0.8-py3-none-any.whl (19 kB)
Collecting zipp>=0.5
  Using cached zipp-3.5.0-py3-none-any.whl (5.7 kB)
Collecting future
  Using cached future-0.18.2-py3-none-any.whl
Collecting scipy
  Using cached scipy-1.5.4-cp36-cp36m-macosx_10_9_x86_64.whl (28.8 MB)
Collecting hyperopt
  Using cached hyperopt-0.2.5-py2.py3-none-any.whl (965 kB)
Collecting pyyaml
  Using cached PyYAML-5.4.1-cp36-cp36m-macosx_10_9_x86_64.whl (249 kB)
Collecting matplotlib
  Using cached matplotlib-3.3.4-cp36-cp36m-macosx_10_9_x86_64.whl (8.5 MB)
Collecting joblib>=0.11
  Using cached joblib-1.0.1-py3-none-any.whl (303 kB)
Collecting threadpoolctl>=2.0.0
  Using cached threadpoolctl-2.2.0-py3-none-any.whl (12 kB)
Collecting networkx>=2.2
  Using cached networkx-2.5.1-py3-none-any.whl (1.6 MB)
Collecting cloudpickle
  Using cached cloudpickle-2.0.0-py3-none-any.whl (25 kB)
Collecting tqdm
  Using cached tqdm-4.62.2-py2.py3-none-any.whl (76 kB)
Collecting decorator<5,>=4.3
  Using cached decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
Collecting pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.3
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Collecting kiwisolver>=1.0.1
  Using cached kiwisolver-1.3.1-cp36-cp36m-macosx_10_9_x86_64.whl (61 kB)
Collecting python-dateutil>=2.1
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting pillow>=6.2.0
  Using cached Pillow-8.3.2-cp36-cp36m-macosx_10_10_x86_64.whl (3.0 MB)
Collecting cycler>=0.10
  Using cached cycler-0.10.0-py2.py3-none-any.whl (6.5 kB)
Collecting pytz>=2017.2
  Using cached pytz-2021.1-py2.py3-none-any.whl (510 kB)
Collecting seaborn
  Using cached seaborn-0.11.2-py3-none-any.whl (292 kB)
Collecting umap-learn>=0.3.10
  Using cached umap_learn-0.5.1-py3-none-any.whl
Collecting anndata>=0.7.4
  Using cached anndata-0.7.6-py3-none-any.whl (127 kB)
Collecting legacy-api-wrap
  Using cached legacy_api_wrap-1.2-py3-none-any.whl (37 kB)
Collecting tables
  Using cached tables-3.6.1-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl (4.4 MB)
Collecting numba>=0.41.0
  Using cached numba-0.53.1-cp36-cp36m-macosx_10_14_x86_64.whl (2.2 MB)
Collecting patsy
  Using cached patsy-0.5.1-py2.py3-none-any.whl (231 kB)
Collecting packaging
  Using cached packaging-21.0-py3-none-any.whl (40 kB)
Collecting natsort
  Using cached natsort-7.1.1-py3-none-any.whl (35 kB)
Collecting statsmodels>=0.10.0rc2
  Using cached statsmodels-0.12.2-cp36-cp36m-macosx_10_15_x86_64.whl (9.5 MB)
Collecting sinfo
  Using cached sinfo-0.3.4-py3-none-any.whl
Collecting xlrd<2.0
  Using cached xlrd-1.2.0-py2.py3-none-any.whl (103 kB)
Collecting llvmlite<0.37,>=0.36.0rc1
  Using cached llvmlite-0.36.0-cp36-cp36m-macosx_10_9_x86_64.whl (18.5 MB)
Collecting pynndescent>=0.5
  Using cached pynndescent-0.5.4-py3-none-any.whl
Collecting get-version>=2.0.4
  Using cached get_version-2.1-py3-none-any.whl (43 kB)
Collecting stdlib-list
  Using cached stdlib_list-0.8.0-py3-none-any.whl (63 kB)
Collecting numexpr>=2.6.2
  Using cached numexpr-2.7.3-cp36-cp36m-macosx_10_9_x86_64.whl (101 kB)
Installing collected packages: urllib3, pyasn1, numpy, idna, charset-normalizer, zipp, typing-extensions, threadpoolctl, six, scipy, rsa, requests, pyasn1-modules, oauthlib, llvmlite, joblib, cachetools, scikit-learn, requests-oauthlib, pytz, python-dateutil, pyparsing, pillow, numba, kiwisolver, importlib-metadata, google-auth, decorator, dataclasses, cycler, cached-property, xlrd, werkzeug, tqdm, tensorboard-plugin-wit, tensorboard-data-server, stdlib-list, pynndescent, protobuf, patsy, pandas, packaging, numexpr, networkx, natsort, matplotlib, markdown, h5py, grpcio, google-auth-oauthlib, get-version, future, cloudpickle, absl-py, wrapt, umap-learn, termcolor, tensorflow-estimator, tensorboard, tables, statsmodels, sinfo, seaborn, pyyaml, opt-einsum, legacy-api-wrap, keras-preprocessing, keras, hyperopt, google-pasta, gast, flatbuffers, clang, astunparse, anndata, tensorflow, scanpy, kopt, dca
Successfully installed absl-py-0.13.0 anndata-0.7.6 astunparse-1.6.3 cached-property-1.5.2 cachetools-4.2.2 charset-normalizer-2.0.5 clang-5.0 cloudpickle-2.0.0 cycler-0.10.0 dataclasses-0.8 dca-0.3.2 decorator-4.4.2 flatbuffers-1.12 future-0.18.2 gast-0.4.0 get-version-2.1 google-auth-1.35.0 google-auth-oauthlib-0.4.6 google-pasta-0.2.0 grpcio-1.40.0 h5py-3.1.0 hyperopt-0.2.5 idna-3.2 importlib-metadata-4.8.1 joblib-1.0.1 keras-2.6.0 keras-preprocessing-1.1.2 kiwisolver-1.3.1 kopt-0.1.0 legacy-api-wrap-1.2 llvmlite-0.36.0 markdown-3.3.4 matplotlib-3.3.4 natsort-7.1.1 networkx-2.5.1 numba-0.53.1 numexpr-2.7.3 numpy-1.19.5 oauthlib-3.1.1 opt-einsum-3.3.0 packaging-21.0 pandas-1.1.5 patsy-0.5.1 pillow-8.3.2 protobuf-3.18.0 pyasn1-0.4.8 pyasn1-modules-0.2.8 pynndescent-0.5.4 pyparsing-2.4.7 python-dateutil-2.8.2 pytz-2021.1 pyyaml-5.4.1 requests-2.26.0 requests-oauthlib-1.3.0 rsa-4.7.2 scanpy-1.7.2 scikit-learn-0.24.2 scipy-1.5.4 seaborn-0.11.2 sinfo-0.3.4 six-1.15.0 statsmodels-0.12.2 stdlib-list-0.8.0 tables-3.6.1 tensorboard-2.6.0 tensorboard-data-server-0.6.1 tensorboard-plugin-wit-1.8.0 tensorflow-2.6.0 tensorflow-estimator-2.6.0 termcolor-1.1.0 threadpoolctl-2.2.0 tqdm-4.62.2 typing-extensions-3.7.4.3 umap-learn-0.5.1 urllib3-1.26.6 werkzeug-2.0.1 wrapt-1.12.1 xlrd-1.2.0 zipp-3.5.0

~ 1m 38s
test ❯ python
Python 3.6.13 |Anaconda, Inc.| (default, Feb 23 2021, 12:58:59)
[GCC Clang 10.0.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
> from dca.api import dca

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/calderwa/opt/anaconda3/envs/test/lib/python3.6/site-packages/dca/api.py", line 15, in <module>
    from .train import train
  File "/Users/calderwa/opt/anaconda3/envs/test/lib/python3.6/site-packages/dca/train.py", line 24, in <module>
    from .network import AE_types
  File "/Users/calderwa/opt/anaconda3/envs/test/lib/python3.6/site-packages/dca/network.py", line 27, in <module>
    from keras.objectives import mean_squared_error
ModuleNotFoundError: No module named 'keras.objectives'
>>>

**Comment by gokceneraslan** [comment]
Hi Alex,

Apparently keras 2.6.0 broke the API significantly, so I set the required keras version for DCA as >=2.4 and <2.6, and made a new release. Please try installing dca 0.3.3 or just downgrading your keras version to 2.4.x. Let me know how it goes.

**Comment by AlexCalderwood** [comment]
Thanks Gokcen, I've tried installing dca 0.3.3, but without much success - I get a new keras related error (see below). Which keras version are you using? and I'll try it with that

```
import pkg_resources
print(f"dca v{pkg_resources.get_distribution('dca').version}")
print(f"keras v{pkg_resources.get_distribution('keras').version}")

from dca.api import dca
```

output:

dca v0.3.3
keras v2.4.3

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-5c8fd6c10c0e> in <module>
      7 print(f"keras v{pkg_resources.get_distribution('keras').version}")
      8 
----> 9 from dca.api import dca

~/.conda/envs/DCA/lib/python3.6/site-packages/dca/api.py in <module>
     13 
     14 from .io import read_dataset, normalize
---> 15 from .train import train
     16 from .network import AE_types
     17 

~/.conda/envs/DCA/lib/python3.6/site-packages/dca/train.py in <module>
     22 
     23 from . import io
---> 24 from .network import AE_types
     25 from .hyper import hyper
     26 

~/.conda/envs/DCA/lib/python3.6/site-packages/dca/network.py in <module>
     21 import scanpy as sc
     22 
---> 23 import keras
     24 from keras.layers import Input, Dense, Dropout, Activation, BatchNormalization, Lambda
     25 from keras.models import Model

~/.conda/envs/DCA/lib/python3.6/site-packages/keras/__init__.py in <module>
     18 from . import callbacks
     19 from . import constraints
---> 20 from . import initializers
     21 from . import metrics
     22 from . import models

~/.conda/envs/DCA/lib/python3.6/site-packages/keras/initializers/__init__.py in <module>
    122 # from ALL_OBJECTS. We make no guarantees as to whether these objects will
    123 # using their correct version.
--> 124 populate_deserializable_objects()
    125 globals().update(LOCAL.ALL_OBJECTS)
    126 

~/.conda/envs/DCA/lib/python3.6/site-packages/keras/initializers/__init__.py in populate_deserializable_objects()
     80     v2_objs = {}
     81     base_cls = initializers_v2.Initializer
---> 82     generic_utils.populate_dict_with_module_objects(
     83         v2_objs,
     84         [initializers_v2],

AttributeError: module 'keras.utils.generic_utils' has no attribute 'populate_dict_with_module_objects'

**Comment by gokceneraslan** [comment]
This was reported at keras repository before https://github.com/keras-team/keras/issues/14632. Do you mind downgrading TF also to 2.4.x? If that fixes it, I will pin the TF version to 2.4 too.

**Comment by AlexCalderwood** [comment]
Ace, that seems to have got it working using 
keras v2.4.3
tf v2.4.0

thanks!

**Comment by gokceneraslan** [comment]
Made a new release, closing this report. Thanks for quick responses :)

**Comment by YichaoOU** [comment]
For anyone who got similar issue, yes, `keras v2.4.3` and `tf v2.4.0` is the solution. 

My whole installation story: I first got `ImportError: cannot import name 'LayerNormalization'`, even though I initially had tensorflow 2.4 installed. I've tried many ways to solve this problem and found upgrading to v2.6 solved it (https://github.com/tensorflow/tensorflow/issues/50004), then I got a `scanpy` issue and later one I found it is the `anndata` problem. After solving all of them, I got this `keras.objectives` issue again. I know previously I have solved it (this is the loop), so I `manually deleted all tensorflow-*` in the site-packages and reinstalled tensorflow 2.4 and keras 2.4.3. And now the DCA run smoothly.

**Comment by zoejuergen** [comment]
> For anyone who got similar issue, yes, `keras v2.4.3` and `tf v2.4.0` is the solution.
> 
> My whole installation story: I first got `ImportError: cannot import name 'LayerNormalization'`, even though I initially had tensorflow 2.4 installed. I've tried many ways to solve this problem and found upgrading to v2.6 solved it ([tensorflow/tensorflow#50004](https://github.com/tensorflow/tensorflow/issues/50004)), then I got a `scanpy` issue and later one I found it is the `anndata` problem. After solving all of them, I got this `keras.objectives` issue again. I know previously I have solved it (this is the loop), so I `manually deleted all tensorflow-*` in the site-packages and reinstalled tensorflow 2.4 and keras 2.4.3. And now the DCA run smoothly.

also important is to use python version='3.7.*|3.8.*|3.9.*'
